README
newsmangler is a basic client for posting to Usenet. The only notable feature is
multiple connection support to efficiently utilize modern bandwidth. 


INSTALLATION
------------
Copy sample.conf to ~/.newsmangler.conf, edit the options as appropriate.


USAGE
-----
Make a directory containing the files you wish to post, the _directory name_ will
be used as the post subject. For example:

test post please ignore/
 \- test.part1.rar
 \- test.part2.rar


docs/CHANGELOG
2012-12-03: * Hopefully fix NZBs being generated with the segments tree in a
              strange order.

2012-04-01: * Fix some issues with the fake Poll implementation on Windows
              systems - apparently someone uses this instead of a good client.
              [by JackDandy]

2012-03-15: * Finish rewrite of how articles are handled:
              - Wrap article information in Article objects so we don't use
                awful magical lists to hand data around.
              - Don't read/encode article data until it's needed. This means
                we no longer read the entire set of files to post into memory
                at startup.
              - Helps lay the framework for tracking which posts failed and
                either retrying them straight away or later.
            * Made sure that copyright notices are up to date.
            * Add *.nzb to .gitignore.
            * Various misc code cleanups.

2012-03-12: * Completely rearrange directory structure to match what is
              expected of a Python app.
            * Improve error handling around connection teardown.
            * Combined BaseMangler and Poster into PostMangler. No real
              reason to have two separate classes.
            * Don't loop until a connection connects before doing anything,
              no posting will be done until one is idle anyway.

2012-03-08: * Remove leecher.py and classes/Leecher.py, there are plenty of
              great options for downloading files already.
            * Add -d/--debug option to actually display debug information.
            * Add some extra debug logging to classes/asyncNNTP.
            * Change the .nzb generation code to use ElementTree, no reason
              at all to do this manually.
            * Update README.

2012-01-13: * Ugly fix for an article that ends with a line consisting of a
              single space/tab character.
            * Fix unquoted ampersands in XML output breaking strict parsers.

2010-11-07: * Change asyncNNTP to extend asyncore.dispatcher channel methods
              properly.

2008-05-09: * Strip whitespace from the newsgroup list to obey RFC1036.

2005-11-11: * Change the Message-ID generation slightly to better match
              "expected" behaviour. (patch by Super-Fly)
            * Fix the =ybegin line so it has the correct line length.
              (patch by Super-Fly)

2005-11-06: * Bump the version to 0.02 to try and track down which version
              people are using.
            * Add the yEncode method we're using to the X-Newsposter header.
            * Require a newer ID string from yenc-freddie for it to be
              recognised.

2005-11-04: * Fix posted files count being incorrect if we end up skipping some
              things.

2005-10-25: * Modify some logic in yEncode_Python to follow the yEnc specs a
              little better. We now escape a leading period on a line, and
              never write more than linelen+1 bytes per line.

2005-10-21: * Add this file!
            * Add a log message detailing which yEncode version we're using.

2005-10-20: * Speed up article list generation a bit.
            * Add -fSUBJECT option for posting an arbitrary list of files
              instead of a directory/directories.
            * Use Psyco to speed up part encoding by 30-35% if it's available.
            * Use my modified yenc module to speed up part encoding by 65-70%
              if it's available.
            * Strip the directory name from the filename field of =ybegin when
              using files mode.
            * Reduce memory thrashing by not continually creating a new string
              for our data buffer. We now just use a pointer value and only
              read a new block of data in once we have exhausted the current
              one.
            * Add FakePoll class, emulates select.poll() on systems that do not
              have it (such as Windows).
            * Add -cCONFIG option to specify a different config file.

2005-10-19: * Print a status message once a second showing our progress.

2005-10-18: * Move some more logging to DEBUG level.
            * Don't set our posting start time until at least one thread is
              connected. This makes our posting speed more accurate.
            * Fix an invalid exception handler in asyncNNTP.
            * Only increment our byte count if we're posting a file, commands
              shouldn't count.

2005-10-17: * Disable our Date: header generator, let the server do it.
            * Modify the SO_SNDBUF of our sockets before we try to connect.

2005-10-14: * Add posting/skip_filenames option.
            * Clarify the comment for posting/default_group.
            * Fix the begin field of our =ybegin lines being off by one. This
              fixes decoding on NZB-O-Matic at least.
            * Write all groups out to the generated .nzb file if posting to
              more than one.


classes/Poster.py
# ---------------------------------------------------------------------------
# $Id: poster.py 3875 2005-10-03 08:19:19Z freddie $
# ---------------------------------------------------------------------------
# Copyright (c) 2005, freddie@madcowdisease.org
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#   * Redistributions of source code must retain the above copyright notice,
#     this list of conditions, and the following disclaimer.
#   * Redistributions in binary form must reproduce the above copyright notice,
#     this list of conditions, and the following disclaimer in the
#     documentation and/or other materials provided with the distribution.
#   * Neither the name of the author of this software nor the name of
#     contributors to this software may be used to endorse or promote products
#     derived from this software without specific prior written consent.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

"""Main class for posting stuff."""

import asyncore
import os
import select
import sys
import time

from cStringIO import StringIO

try:
    import xml.etree.cElementTree as ET
except:
    import xml.etree.ElementTree as ET

from classes import asyncNNTP
from classes import yEnc
from classes.BaseMangler import BaseMangler
from classes.Common import *

# ---------------------------------------------------------------------------

class Poster(BaseMangler):
	def __init__(self, conf, debug=False):
		BaseMangler.__init__(self, conf, debug)
		
		self.conf['posting']['skip_filenames'] = self.conf['posting'].get('skip_filenames', '').split()
		
		self._articles = []
		self._files = {}
		self._msgids = {}
		
		self._current_dir = None
		self.newsgroup = None
		self.post_title = None
		self._par2_queue = []
		
		# Some sort of useful logging junk about what yEncode we're using
		self.logger.info('Using %s module for yEnc encoding.', yEnc.yEncMode())
	
	def post(self, newsgroup, postme, post_title=None):
		self.newsgroup = newsgroup
		self.post_title = post_title
		
		# Generate the list of articles we need to post
		self.generate_article_list(postme)
		
		# If we have no valid articles, bail
		if not self._articles:
			self.logger.warning('No valid articles to post!')
			return
		
		# Connect!
		self.connect()
		
		# Slight speedup
		_poll = self.poll
		_sleep = time.sleep
		_time = time.time
		
		# Wait until at least one connection is ready
		last_stuff = _time()
		while 1:
			now = _time()
			_poll()
			
			if self._idle:
				break
			
			if now - last_stuff >= 1:
				last_stuff = now
				for conn in self._conns:
					if conn.state == asyncNNTP.STATE_DISCONNECTED and now >= conn.reconnect_at:
						conn.do_connect()
			
			_sleep(0.01)
		
		# And loop
		self._bytes = 0
		start = _time()
		
		self.logger.info('Posting %d article(s)...', len(self._articles))
		
		while 1:
			now = _time()
			
			# Poll our sockets for events
			_poll()
			
			# Possibly post some more parts now
			while self._idle and self._articles:
				article = self._articles.pop(0)
				postfile = StringIO()
				self.build_article(postfile, article)
				
				conn = self._idle.pop(0)
				conn.post_article(postfile)
			
			# Do some stuff roughly once per second
			if now - last_stuff >= 1:
				last_stuff = now
				
				for conn in self._conns:
					if conn.state == asyncNNTP.STATE_DISCONNECTED and now >= conn.reconnect_at:
						conn.do_connect()
				
				if self._bytes:
					interval = time.time() - start
					speed = self._bytes / interval / 1024
					left = len(self._articles) + (len(self._conns) - len(self._idle))
					print '%d articles remaining - %.1fKB/s     \r' % (left, speed),
					sys.stdout.flush()
			
			# All done?
			if self._articles == [] and len(self._idle) == self.conf['server']['connections']:
				interval = time.time() - start
				speed = self._bytes / interval / 1024
				self.logger.info('Posting complete - %d bytes in %.2fs (%.1fKB/s)',
					self._bytes, interval, speed)
				
				# If we have some msgids left over, we might have to generate
				# a .NZB
				if self.conf['posting']['generate_nzbs'] and self._msgids:
					self.generate_nzb()
				
				return
			
			# And sleep for a bit to try and cut CPU chompage
			_sleep(0.01)
	
	# -----------------------------------------------------------------------
	# Generate the list of articles we need to post
	def generate_article_list(self, postme):
		# "files" mode is just one lot of files
		if self.post_title:
			self._gal_files(self.post_title, postme)
		# "dirs" mode could be a whole bunch
		else:
			for dirname in postme:
				if dirname.endswith(os.sep):
					dirname = dirname[:-len(os.sep)]
				if not dirname:
					continue
				
				self._gal_files(os.path.basename(dirname), os.listdir(dirname), basepath=dirname)
		
		# Debug junk
		if 0:
			for article in self._articles:
				print article[1]
	
	# Do the heavy lifting for generate_article_list
	def _gal_files(self, post_title, files, basepath=''):
		article_size = self.conf['posting']['article_size']
		
		goodfiles = []
		for filename in files:
			filepath = os.path.abspath(os.path.join(basepath, filename))
			
			# Skip non-files and empty files
			if not os.path.isfile(filepath):
				continue
			if filename in self.conf['posting']['skip_filenames']:
				continue
			filesize = os.path.getsize(filepath)
			if not filesize:
				continue
			
			goodfiles.append((filename, filepath, filesize))
		goodfiles.sort()
		
		n = 1
		for filename, filepath, filesize in goodfiles:
			full, partial = divmod(filesize, article_size)
			if partial:
				parts = full + 1
			else:
				parts = full
			
			# Build a subject
			real_filename = os.path.split(filename)[1]
			
			temp = '%%0%sd' % (len(str(len(files))))
			filenum = temp % (n)
			temp = '%%0%sd' % (len(str(parts)))
			subject = '%s [%s/%d] - "%s" yEnc (%s/%d)' % (
				post_title, filenum, len(goodfiles), real_filename, temp, parts
			)
			
			if self.conf['posting']['subject_prefix']:
				subject = '%s %s' % (self.conf['posting']['subject_prefix'], subject)
			
			# Now make up our parts
			fileinfo = {
				'dirname': post_title,
				'filename': real_filename,
				'filepath': filepath,
				'filesize': filesize,
				'parts': parts,
			}
			
			for i in range(parts):
				article = [fileinfo, subject, i+1]
				self._articles.append(article)
			
			n += 1
	
	# -----------------------------------------------------------------------
	# Build an article for posting.
	def build_article(self, postfile, article):
		(fileinfo, subject, partnum) = article
		
		# Read the chunk of data from the file
		f = self._files.get(fileinfo['filepath'], None)
		if f is None:
			self._files[fileinfo['filepath']] = f = open(fileinfo['filepath'], 'rb')
		
		begin = f.tell()
		data = f.read(self.conf['posting']['article_size'])
		end = f.tell()
		
		# If that was the last part, close the file and throw it away
		if partnum == fileinfo['parts']:
			self._files[fileinfo['filepath']].close()
			del self._files[fileinfo['filepath']]
		
		# Basic headers
		line = 'From: %s\r\n' % (self.conf['posting']['from'])
		postfile.write(line)
		
		line = 'Newsgroups: %s\r\n' % (self.newsgroup)
		postfile.write(line)
		
		#line = time.strftime('Date: %a, %d %b %Y %H:%M:%S GMT\r\n', time.gmtime())
		#postfile.write(line)
		
		subj = subject % (partnum)
		line = 'Subject: %s\r\n' % (subj)
		postfile.write(line)
		
		msgid = '%.5f.%d@%s' % (time.time(), partnum, self.conf['server']['hostname'])
		line = 'Message-ID: <%s>\r\n' % (msgid)
		postfile.write(line)
		
		line = 'X-Newsposter: newsmangler %s (%s) - http://www.madcowdisease.org/mcd/newsmangler\r\n' % (
			NM_VERSION, yEnc.yEncMode())
		postfile.write(line)
		
		postfile.write('\r\n')
		
		# yEnc start
		line = '=ybegin part=%d total=%d line=128 size=%d name=%s\r\n' % (
			partnum, fileinfo['parts'], fileinfo['filesize'], fileinfo['filename']
		)
		postfile.write(line)
		line = '=ypart begin=%d end=%d\r\n' % (begin+1, end)
		postfile.write(line)
		
		# yEnc data
		partcrc = yEnc.yEncode(postfile, data)
		
		# yEnc end
		line = '=yend size=%d part=%d pcrc32=%s\r\n' % (end-begin, partnum, partcrc)
		postfile.write(line)
		
		# And done writing for now
		postfile.write('.\r\n')
		article_size = postfile.tell()
		postfile.seek(0, 0)
		
		# Maybe remember the msgid for later
		if self.conf['posting']['generate_nzbs']:
			if self._current_dir != fileinfo['dirname']:
				if self._msgids:
					self.generate_nzb()
					self._msgids = {}
				
				self._current_dir = fileinfo['dirname']
			
			subj = subject % (1)
			if subj not in self._msgids:
				self._msgids[subj] = [int(time.time())]
			self._msgids[subj].append((msgid, article_size))
	
	# -----------------------------------------------------------------------
	# Generate a .NZB file!
	def generate_nzb(self):
		filename = 'newsmangler_%s.nzb' % (SafeFilename(self._current_dir))
		
		self.logger.info('Begin generation of %s', filename)

		gentime = time.strftime('%Y-%m-%d %H:%M:%S UTC', time.gmtime())

		root = ET.Element('nzb')
		root.append(ET.Comment('Generated by newsmangler v%s at %s' % (NM_VERSION, gentime)))

		for subject, msgids in self._msgids.items():
			posttime = msgids.pop(0)

			print posttime

			# file
			f = ET.SubElement(root, 'file',
				{
					'poster': self.conf['posting']['from'],
					'date': str(posttime),
					'subject': subject,
				}
			)

			# newsgroups
			groups = ET.SubElement(f, 'groups')
			for newsgroup in self.newsgroup.split(','):
				group = ET.SubElement(groups, 'group')
				group.text = newsgroup

			# segments
			segments = ET.SubElement(f, 'segments')
			for i, (msgid, article_size) in enumerate(msgids):
				segment = ET.SubElement(segments, 'segment',
					{
						'bytes': str(article_size),
						'number': str(i + 1),
					}
				)
				segment.text = str(msgid)

		with open(filename, 'w') as nzbfile:
			ET.ElementTree(root).write(nzbfile, xml_declaration=True)

		self.logger.info('End generation of %s', filename)

# ---------------------------------------------------------------------------
